SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialFunctionalTest.cpp

```diff
@@ -0,0 +1,547 @@
+		if (FlowController->ControllerType == ControllerType && FlowController->ControllerInstanceId == InstanceId)
+		{
+			return FlowController;
+		}
+	}
+	return nullptr;
+}
+
+void ASpatialFunctionalTest::CrossServerNotifyStepFinished_Implementation(ASpatialFunctionalTestFlowController* FlowController)
+{
+	if (CurrentStepIndex < 0)
+	{
+		return;
+	}
+
+	const FString FLowControllerDisplayName = FlowController->GetDisplayName();
+	
+	UE_LOG(LogSpatialFunctionalTest, Display, TEXT("%s finished Step"), *FLowControllerDisplayName);
+	
+	if (FlowControllersExecutingStep.RemoveSwap(FlowController) == 0)
+	{
+		FString ErrorMsg = FString::Printf(TEXT("%s was not in list of workers executing"), *FLowControllerDisplayName);
+		checkf(false, TEXT("%s"), *ErrorMsg);
+		FinishTest(EFunctionalTestResult::Error, ErrorMsg);
+	}
+}
+
+void ASpatialFunctionalTest::OnReplicated_CurrentStepIndex()
+{
+	if (CurrentStepIndex == SPATIAL_FUNCTIONAL_TEST_FINISHED)
+	{
+		//test finished
+		if(StartTime > 0)
+		{
+			//if we ever started in first place
+			ASpatialFunctionalTestFlowController* AuxLocalFlowController = GetLocalFlowController();
+			if (AuxLocalFlowController != nullptr)
+			{
+				AuxLocalFlowController->OnTestFinished();
+			}
+		}
+		if (!HasAuthority()) // Authority already does this on Super::FinishTest
+		{
+			NotifyTestFinishedObserver();
+		}
+	}
+}
+
+void ASpatialFunctionalTest::StartServerFlowControllerSpawn()
+{
+	if (!bReadyToSpawnServerControllers)
+	{
+		return;
+	}
+
+	if (FlowControllerActorClass.Get() != nullptr)
+	{
+		FlowControllerSpawner.ModifyFlowControllerClassToSpawn(FlowControllerActorClass);
+	}
+
+	FlowControllerSpawner.SpawnServerFlowController();
+}
+
+void ASpatialFunctionalTest::SetupClientPlayerRegistrationFlow()
+{
+	GetWorld()->AddOnActorSpawnedHandler(FOnActorSpawned::FDelegate::CreateLambda(
+		[this](AActor* Spawned)
+		{
+			if (APlayerController* PlayerController = Cast<APlayerController>(Spawned))
+			{
+				this->FlowControllerSpawner.SpawnClientFlowController(PlayerController);
+			}
+		}
+	));
+}
+
+void ASpatialFunctionalTest::CrossServerRegisterAutoDestroyActor_Implementation(AActor* ActorToAutoDestroy)
+{
+	RegisterAutoDestroyActor(ActorToAutoDestroy);
+}
+
+void ASpatialFunctionalTest::ServerRegisterAutoDestroyActor_Implementation(AActor* ActorToAutoDestroy)
+{
+	CrossServerRegisterAutoDestroyActor(ActorToAutoDestroy);
+}
+
+void ASpatialFunctionalTest::MulticastAutoDestroyActors_Implementation(const TArray<AActor*>& ActorsToDestroy)
+{
+	FString DisplayName = LocalFlowController ? LocalFlowController->GetDisplayName() : TEXT("UNKNOWN");
+	if (!HasAuthority()) // Authority already handles it in Super::FinishTest
+	{
+		for (AActor* Actor : ActorsToDestroy)
+		{
+			if (IsValid(Actor))
+			{				
+				UE_LOG(LogSpatialFunctionalTest, Display, TEXT("%s trying to delete actor: %s ; result now would be: %s"), *DisplayName, *Actor->GetName(), Actor->Role == ROLE_Authority ? TEXT("SUCCESS") : TEXT("FAILURE"));
```

Comment:
Note: these logs were added for debugging and future replication of bug GSE-1114